flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/rules/logical/JoinRemoveRulesTest.scala

```diff
@@ -0,0 +1,311 @@
+    util.verifyRelPlan(sqlQuery)
+  }
+
+  @Test
+  def testAggregateJoinRemove9(): Unit = {
+    val sqlQuery =
+      s"""
+         |SELECT distinct e.deptno, d2.name
+         |FROM emp e
+         |LEFT JOIN dept d1 ON e.deptno = d1.deptno
+         |LEFT JOIN dept d2 ON e.deptno = d2.deptno
+       """.stripMargin
+    util.verifyRelPlan(sqlQuery)
+  }
+
+  @Test
+  def testAggregateJoinRemove10(): Unit = {
+    val sqlQuery =
+      s"""
+         |SELECT e.deptno, COUNT(DISTINCT d1.name, d2.name)
+         |FROM emp e
+         |LEFT JOIN dept d1 ON e.deptno = d1.deptno
+         |LEFT JOIN dept d2 ON e.deptno = d2.deptno
+         |GROUP BY e.deptno
+       """.stripMargin
+    util.verifyRelPlan(sqlQuery)
+  }
+
+  @Test
+  def testProjectJoinRemove1(): Unit = {
+    val sqlQuery =
+      s"""
+         |SELECT e.deptno, d2.deptno
+         |FROM emp e
+         |LEFT JOIN dept d1 ON e.deptno = d1.deptno
+         |LEFT JOIN dept d2 ON e.deptno = d2.deptno
+       """.stripMargin
+    util.verifyRelPlan(sqlQuery)
+  }
+
+  @Test
+  def testProjectJoinRemove2(): Unit = {
+    val sqlQuery =
+      s"""
+         |SELECT e.deptno, d1.deptno
+         |FROM emp e
+         |LEFT JOIN dept d1 ON e.deptno = d1.deptno
+         |LEFT JOIN dept d2 ON e.deptno = d2.deptno
+       """.stripMargin
+    util.verifyRelPlan(sqlQuery)
+  }
+
+  @Test
+  def testProjectJoinRemove3(): Unit = {
+    val sqlQuery =
+      s"""
+         |SELECT e1.deptno, d.deptno
+         |FROM emp e1
+         |LEFT JOIN emp e2 ON e1.deptno = e2.deptno
+         |LEFT JOIN dept d ON e1.deptno = d.deptno
+       """.stripMargin
+    util.verifyRelPlan(sqlQuery)
+  }
+
+  @Test
+  def testProjectJoinRemove4(): Unit = {
+    val sqlQuery =
+      s"""
+         |SELECT e.deptno
+         |FROM emp e
+         |LEFT JOIN dept d ON e.deptno = d.deptno
+       """.stripMargin
+    util.verifyRelPlan(sqlQuery)
+  }
+
+  @Test
+  def testProjectJoinRemove5(): Unit = {
+    val sqlQuery =
+      s"""
+         |SELECT e1.deptno
+         |FROM emp e1
+         |LEFT JOIN emp e2 ON e1.deptno = e2.deptno
+       """.stripMargin
+    util.verifyRelPlan(sqlQuery)
+  }
+
+  @Test
+  def testProjectJoinRemove6(): Unit = {
+    val sqlQuery =
+      s"""
+         |SELECT e.deptno, d.name
+         |FROM emp e
+         |LEFT JOIN dept d ON e.deptno = d.deptno
+       """.stripMargin
+    util.verifyRelPlan(sqlQuery)
+  }
+
+  @Test
+  def testProjectJoinRemove7(): Unit = {
+    val sqlQuery =
+      s"""
+         |SELECT e.deptno
+         |FROM dept d
+         |RIGHT JOIN emp e ON e.deptno = d.deptno
+       """.stripMargin
+    util.verifyRelPlan(sqlQuery)
+  }
+
+  @Test
+  def testProjectJoinRemove8(): Unit = {
+    val sqlQuery =
+      s"""
+         |SELECT e2.deptno
+         |FROM emp e1
+         |LEFT JOIN emp e2 ON e1.deptno = e2.deptno
+       """.stripMargin
+    util.verifyRelPlan(sqlQuery)
+  }
+
+  @Test
+  def testProjectJoinRemove9(): Unit = {
```

Comment:
If I understand correctly, all these `testProjectJoinRemove` fail to remove the join because they don't meet the require that the projected col is unique on the corresponding table. Please take a look.